In [3]:
import pandas as pd

## Pandas Dataframes

In previous years we ran this course using R, however this year we will continue to work with Python. As you have learned, the most common library for manipulating structured data in Python is called `pandas`. This notebook covers a quick refresher and tour, using some of the data we will work with later in this course.

There are a lot of pandas resources available, eg [cheatsheets](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf).

The data that we have just loaded comes from the file containing the Arthurian manuscripts metadata. Each row represents a single manuscript and the columns describe different properties of each of these books (or at least, what remains of them).

First, let's load the data from a CSV file on disk.

In [4]:
df = pd.read_csv("../../datasets/arthur/manuscripts.csv", index_col=0)
df

signature  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...   
2          Berlin, Staatsbibliothek, germ. fol. 923, 38   
3                          Linz, Landesarchiv, 96/IV/54   
4           Brussels, Koninklijke Bibliotheek, II 115,2   
5              Munich, Staatsbibliothek, cgm. 5249 / 3b   
...                                                 ...   
1162                Orléans, Bibliothèque Municipale, -   
1163              Freiburg, Universitätsbibliothek, 513   
1164   Paris, Bibliothèque Nationale de France, fr. 340   
1165        Heidelberg, Universitätsbibliothek, Cpg 397   
1166     Florence, Biblioteca Laurenziana, Plut. 77, 23   

                                             repository     script  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...  textualis   
2                              Berlin, Staatsbibliothek        NaN   
3                                    Linz, Landesarchiv        NaN   
4                     Brussels, Koninklijke Bibliotheek  textualis   
5                              Munich, Staatsbibliothek  textualis   
...                                                 ...        ...   
1162                   Orléans, Bibliothèque Municipale        NaN   
1163                   Freiburg, Universitätsbibliothek        NaN   
1164            Paris, Bibliothèque Nationale de France        NaN   
1165                 Heidelberg, Universitätsbibliothek  textualis   
1166                   Florence, Biblioteca Laurenziana        NaN   

      text-height  text-width  text-src  illustrations  date_min  date_max  \
1           220.0       164.0  original              0    1320.0    1325.0   
2             NaN         NaN       NaN              0    1300.0    1400.0   
3           212.0       163.0  original              0    1390.0    1410.0   
4           185.0       132.0  original              0    1340.0    1360.0   
5           195.0       127.0  original              0    1275.0    1300.0   
...           ...         ...       ...            ...       ...       ...   
1162          NaN         NaN       NaN              0    1200.0    1300.0   
1163        150.0       115.0  original              0    1290.0    1310.0   
1164          NaN         NaN       NaN             76    1400.0    1425.0   
1165        165.0       100.0  original              0    1225.0    1250.0   
1166          NaN         NaN       NaN              0       NaN       NaN   

      leaf-height  leaf-width  leaf-src  number-lines   material  \
1           308.0       215.0   current          32.0  parchment   
2             NaN         NaN       NaN           NaN  parchment   
3           280.0       200.0   current          31.0      paper   
4           230.0       155.0  original          39.0  parchment   
5           232.0       160.0   current          42.0  parchment   
...           ...         ...       ...           ...        ...   
1162        150.0       105.0   current          30.0  parchment   
1163        160.0       140.0   current           NaN  parchment   
1164        420.0       300.0   current          53.0  parchment   
1165        198.0       122.0   current          26.0  parchment   
1166          NaN         NaN       NaN           NaN      paper   

     physical-type               country        region  columns  
1         fragment                   NaN           NaN      2.0  
2         fragment               Germany        Bayern      NaN  
3            codex                   NaN           NaN      2.0  
4         fragment  Southern Netherlands      Flanders      2.0  
5         fragment                   NaN           NaN      2.0  
...            ...                   ...           ...      ...  
1162      fragment                   NaN           NaN      1.0  
1163      fragment                   NaN           NaN      2.0  
1164         codex                   NaN           NaN      3.0  
1165         codex               Germany  Rheinfranken      1.0  
1166         cod

We can use the `shape` property to find the dimensions of the `DataFrame` object (rows x columns). Remember that Python *properties* are accessed as bare words (`df.shape` not `df.shape()`), whereas *methods* must have parentheses, even when no arguments are given.

In [5]:
df.shape

(1166, 18)

There are many other useful properties. We can look at the `columns`, and later we will make a lot of use of the `index`

In [6]:
df.columns

Index(['signature', 'repository', 'script', 'text-height', 'text-width',
       'text-src', 'illustrations', 'date_min', 'date_max', 'leaf-height',
       'leaf-width', 'leaf-src', 'number-lines', 'material', 'physical-type',
       'country', 'region', 'columns'],
      dtype='object')

### Basic Methods

We can take a quick look at the DataFrame using the `head()` and `tail()` methods, but also note how Jupyter notebooks will already return a basic view of the DataFrame as long as it is the last thing evaluated in a cell, so mostly we will use that 'shortcut' below.

In [7]:
df.head(3)

signature  \
1  Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...   
2       Berlin, Staatsbibliothek, germ. fol. 923, 38   
3                       Linz, Landesarchiv, 96/IV/54   

                                          repository     script  text-height  \
1  Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...  textualis        220.0   
2                           Berlin, Staatsbibliothek        NaN          NaN   
3                                 Linz, Landesarchiv        NaN        212.0   

   text-width  text-src  illustrations  date_min  date_max  leaf-height  \
1       164.0  original              0    1320.0    1325.0        308.0   
2         NaN       NaN              0    1300.0    1400.0          NaN   
3       163.0  original              0    1390.0    1410.0        280.0   

   leaf-width leaf-src  number-lines   material physical-type  country  \
1       215.0  current          32.0  parchment      fragment      NaN   
2         NaN      NaN           NaN  parchment      fragment  Germany   
3       200.0  current          31.0      paper         codex      NaN   

   region  columns  
1     NaN      2.0  
2  Bayern      NaN  
3     NaN      2.0

In [8]:
df.tail(2)

signature  \
1165     Heidelberg, Universitätsbibliothek, Cpg 397   
1166  Florence, Biblioteca Laurenziana, Plut. 77, 23   

                              repository     script  text-height  text-width  \
1165  Heidelberg, Universitätsbibliothek  textualis        165.0       100.0   
1166    Florence, Biblioteca Laurenziana        NaN          NaN         NaN   

      text-src  illustrations  date_min  date_max  leaf-height  leaf-width  \
1165  original              0    1225.0    1250.0        198.0       122.0   
1166       NaN              0       NaN       NaN          NaN         NaN   

     leaf-src  number-lines   material physical-type  country        region  \
1165  current          26.0  parchment         codex  Germany  Rheinfranken   
1166      NaN           NaN      paper         codex    Italy           NaN   

      columns  
1165      1.0  
1166      NaN

In [9]:
# the last object evaluated in a cell is outputted by Jupyter

df

signature  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...   
2          Berlin, Staatsbibliothek, germ. fol. 923, 38   
3                          Linz, Landesarchiv, 96/IV/54   
4           Brussels, Koninklijke Bibliotheek, II 115,2   
5              Munich, Staatsbibliothek, cgm. 5249 / 3b   
...                                                 ...   
1162                Orléans, Bibliothèque Municipale, -   
1163              Freiburg, Universitätsbibliothek, 513   
1164   Paris, Bibliothèque Nationale de France, fr. 340   
1165        Heidelberg, Universitätsbibliothek, Cpg 397   
1166     Florence, Biblioteca Laurenziana, Plut. 77, 23   

                                             repository     script  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...  textualis   
2                              Berlin, Staatsbibliothek        NaN   
3                                    Linz, Landesarchiv        NaN   
4                     Brussels, Koninklijke Bibliotheek  textualis   
5                              Munich, Staatsbibliothek  textualis   
...                                                 ...        ...   
1162                   Orléans, Bibliothèque Municipale        NaN   
1163                   Freiburg, Universitätsbibliothek        NaN   
1164            Paris, Bibliothèque Nationale de France        NaN   
1165                 Heidelberg, Universitätsbibliothek  textualis   
1166                   Florence, Biblioteca Laurenziana        NaN   

      text-height  text-width  text-src  illustrations  date_min  date_max  \
1           220.0       164.0  original              0    1320.0    1325.0   
2             NaN         NaN       NaN              0    1300.0    1400.0   
3           212.0       163.0  original              0    1390.0    1410.0   
4           185.0       132.0  original              0    1340.0    1360.0   
5           195.0       127.0  original              0    1275.0    1300.0   
...           ...         ...       ...            ...       ...       ...   
1162          NaN         NaN       NaN              0    1200.0    1300.0   
1163        150.0       115.0  original              0    1290.0    1310.0   
1164          NaN         NaN       NaN             76    1400.0    1425.0   
1165        165.0       100.0  original              0    1225.0    1250.0   
1166          NaN         NaN       NaN              0       NaN       NaN   

      leaf-height  leaf-width  leaf-src  number-lines   material  \
1           308.0       215.0   current          32.0  parchment   
2             NaN         NaN       NaN           NaN  parchment   
3           280.0       200.0   current          31.0      paper   
4           230.0       155.0  original          39.0  parchment   
5           232.0       160.0   current          42.0  parchment   
...           ...         ...       ...           ...        ...   
1162        150.0       105.0   current          30.0  parchment   
1163        160.0       140.0   current           NaN  parchment   
1164        420.0       300.0   current          53.0  parchment   
1165        198.0       122.0   current          26.0  parchment   
1166          NaN         NaN       NaN           NaN      paper   

     physical-type               country        region  columns  
1         fragment                   NaN           NaN      2.0  
2         fragment               Germany        Bayern      NaN  
3            codex                   NaN           NaN      2.0  
4         fragment  Southern Netherlands      Flanders      2.0  
5         fragment                   NaN           NaN      2.0  
...            ...                   ...           ...      ...  
1162      fragment                   NaN           NaN      1.0  
1163      fragment                   NaN           NaN      2.0  
1164         codex                   NaN           NaN      3.0  
1165         codex               Germany  Rheinfranken      1.0  
1166         cod

If we access an individual column we will get a `Series`. This can also be done with a magic method that treats the column name like a property. The two calls below are mostly equivalent (sometimes the longer string subscript access is required)

In [10]:
df.script

1       textualis
2             NaN
3             NaN
4       textualis
5       textualis
          ...    
1162          NaN
1163          NaN
1164          NaN
1165    textualis
1166          NaN
Name: script, Length: 1166, dtype: object

In [ ]:
df["script"]

Most methods on a DataFrame will return a new DataFrame. Here we pull out the width and height of the pages (leaves) for those manuscripts where the data is recorded. Since that method returns a `DataFrame` we can immediately *chain* another method, `dropna()`, to drop all rows where either value is NA.

NOTE CAREFULLY: The indices for the rows *do not change* (unless we force it to by using a different method). That means our indices are no longer sequential. This is an advantage!

In [12]:
dimensions = df[["leaf-height", "leaf-width"]].dropna()
dimensions.index

Index([   1,    3,    4,    5,    9,   12,   15,   16,   17,   21,
       ...
       1154, 1155, 1157, 1158, 1159, 1160, 1162, 1163, 1164, 1165],
      dtype='int64', length=727)

In [14]:
idx = dimensions.index

In [17]:
type(idx)

pandas.core.indexes.base.Index

In [11]:
df[["leaf-height", "leaf-width"]]

leaf-height  leaf-width
1           308.0       215.0
2             NaN         NaN
3           280.0       200.0
4           230.0       155.0
5           232.0       160.0
...           ...         ...
1162        150.0       105.0
1163        160.0       140.0
1164        420.0       300.0
1165        198.0       122.0
1166          NaN         NaN

[1166 rows x 2 columns]

Here we can pull the rows from the original dataframe, but ONLY at the index locations (`.loc`) specified in the `dimensions` dataframe. Note how we now have a dataframe with only 727 entries.

In [15]:
df.loc[idx]

signature  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...   
3                          Linz, Landesarchiv, 96/IV/54   
4           Brussels, Koninklijke Bibliotheek, II 115,2   
5              Munich, Staatsbibliothek, cgm. 5249 / 3b   
9     Escorial, Real Biblioteca del Monasterio, M. I...   
...                                                 ...   
1160          Marburg, Staatsarchiv, Best. 147 Hr 1, 16   
1162                Orléans, Bibliothèque Municipale, -   
1163              Freiburg, Universitätsbibliothek, 513   
1164   Paris, Bibliothèque Nationale de France, fr. 340   
1165        Heidelberg, Universitätsbibliothek, Cpg 397   

                                             repository     script  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...  textualis   
3                                    Linz, Landesarchiv        NaN   
4                     Brussels, Koninklijke Bibliotheek  textualis   
5                              Munich, Staatsbibliothek  textualis   
9              Escorial, Real Biblioteca del Monasterio  textualis   
...                                                 ...        ...   
1160                              Marburg, Staatsarchiv        NaN   
1162                   Orléans, Bibliothèque Municipale        NaN   
1163                   Freiburg, Universitätsbibliothek        NaN   
1164            Paris, Bibliothèque Nationale de France        NaN   
1165                 Heidelberg, Universitätsbibliothek  textualis   

      text-height  text-width  text-src  illustrations  date_min  date_max  \
1           220.0       164.0  original              0    1320.0    1325.0   
3           212.0       163.0  original              0    1390.0    1410.0   
4           185.0       132.0  original              0    1340.0    1360.0   
5           195.0       127.0  original              0    1275.0    1300.0   
9           224.0       128.0  original              0    1275.0    1300.0   
...           ...         ...       ...            ...       ...       ...   
1160        250.0       190.0  original              0    1200.0    1300.0   
1162          NaN         NaN       NaN              0    1200.0    1300.0   
1163        150.0       115.0  original              0    1290.0    1310.0   
1164          NaN         NaN       NaN             76    1400.0    1425.0   
1165        165.0       100.0  original              0    1225.0    1250.0   

      leaf-height  leaf-width  leaf-src  number-lines   material  \
1           308.0       215.0   current          32.0  parchment   
3           280.0       200.0   current          31.0      paper   
4           230.0       155.0  original          39.0  parchment   
5           232.0       160.0   current          42.0  parchment   
9           260.0       128.0   current          45.0  parchment   
...           ...         ...       ...           ...        ...   
1160         95.0       160.0  original          43.0  parchment   
1162        150.0       105.0   current          30.0  parchment   
1163        160.0       140.0   current           NaN  parchment   
1164        420.0       300.0   current          53.0  parchment   
1165        198.0       122.0   current          26.0  parchment   

     physical-type               country        region  columns  
1         fragment                   NaN           NaN      2.0  
3            codex                   NaN           NaN      2.0  
4         fragment  Southern Netherlands      Flanders      2.0  
5         fragment                   NaN           NaN      2.0  
9            codex                France      Normandy      2.0  
...            ...                   ...           ...      ...  
1160      fragment                   NaN           NaN      2.0  
1162      fragment                   NaN           NaN      1.0  
1163      fragment                   NaN           NaN      2.0  
1164         codex                   NaN           NaN      3.0  
1165         cod

### Operating on Columns

We can insert new `Series` objects (or actually even normal lists, or numpy arrays, etc) into the dataframe as columns. Here, we create one column using a vectorized boolean comparison, and another one which we create by multiplying two existing columns. Once again note how these are *vector* operations -- `ColA * ColB` multiplies the matching values in every row, all in one statement.

Note how in the automatic view the reported shape is increasing by one column each time....

In [19]:
(df.illustrations > 0)

1       False
2       False
3       False
4       False
5       False
        ...  
1162    False
1163    False
1164     True
1165    False
1166    False
Name: illustrations, Length: 1166, dtype: bool

In [18]:
# insert takes index-to-insert, column-name, Series as positional arguments

df.insert(3, "illustrated", df.illustrations > 0)
df

signature  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...   
2          Berlin, Staatsbibliothek, germ. fol. 923, 38   
3                          Linz, Landesarchiv, 96/IV/54   
4           Brussels, Koninklijke Bibliotheek, II 115,2   
5              Munich, Staatsbibliothek, cgm. 5249 / 3b   
...                                                 ...   
1162                Orléans, Bibliothèque Municipale, -   
1163              Freiburg, Universitätsbibliothek, 513   
1164   Paris, Bibliothèque Nationale de France, fr. 340   
1165        Heidelberg, Universitätsbibliothek, Cpg 397   
1166     Florence, Biblioteca Laurenziana, Plut. 77, 23   

                                             repository     script  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...  textualis   
2                              Berlin, Staatsbibliothek        NaN   
3                                    Linz, Landesarchiv        NaN   
4                     Brussels, Koninklijke Bibliotheek  textualis   
5                              Munich, Staatsbibliothek  textualis   
...                                                 ...        ...   
1162                   Orléans, Bibliothèque Municipale        NaN   
1163                   Freiburg, Universitätsbibliothek        NaN   
1164            Paris, Bibliothèque Nationale de France        NaN   
1165                 Heidelberg, Universitätsbibliothek  textualis   
1166                   Florence, Biblioteca Laurenziana        NaN   

      illustrated  text-height  text-width  text-src  illustrations  date_min  \
1           False        220.0       164.0  original              0    1320.0   
2           False          NaN         NaN       NaN              0    1300.0   
3           False        212.0       163.0  original              0    1390.0   
4           False        185.0       132.0  original              0    1340.0   
5           False        195.0       127.0  original              0    1275.0   
...           ...          ...         ...       ...            ...       ...   
1162        False          NaN         NaN       NaN              0    1200.0   
1163        False        150.0       115.0  original              0    1290.0   
1164         True          NaN         NaN       NaN             76    1400.0   
1165        False        165.0       100.0  original              0    1225.0   
1166        False          NaN         NaN       NaN              0       NaN   

      date_max  leaf-height  leaf-width  leaf-src  number-lines   material  \
1       1325.0        308.0       215.0   current          32.0  parchment   
2       1400.0          NaN         NaN       NaN           NaN  parchment   
3       1410.0        280.0       200.0   current          31.0      paper   
4       1360.0        230.0       155.0  original          39.0  parchment   
5       1300.0        232.0       160.0   current          42.0  parchment   
...        ...          ...         ...       ...           ...        ...   
1162    1300.0        150.0       105.0   current          30.0  parchment   
1163    1310.0        160.0       140.0   current           NaN  parchment   
1164    1425.0        420.0       300.0   current          53.0  parchment   
1165    1250.0        198.0       122.0   current          26.0  parchment   
1166       NaN          NaN         NaN       NaN           NaN      paper   

     physical-type               country        region  columns  
1         fragment                   NaN           NaN      2.0  
2         fragment               Germany        Bayern      NaN  
3            codex                   NaN           NaN      2.0  
4         fragment  Southern Netherlands      Flanders      2.0  
5         fragment                   NaN           NaN      2.0  
...            ...                   ...           ...      ...  
1162      fragment                   NaN           NaN      1.0  
1163      fragment                   NaN           NaN    

In [27]:
df.insert(4, "surface", df["leaf-height"] * df["leaf-width"])
df

signature  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...   
2          Berlin, Staatsbibliothek, germ. fol. 923, 38   
3                          Linz, Landesarchiv, 96/IV/54   
4           Brussels, Koninklijke Bibliotheek, II 115,2   
5              Munich, Staatsbibliothek, cgm. 5249 / 3b   
...                                                 ...   
1162                Orléans, Bibliothèque Municipale, -   
1163              Freiburg, Universitätsbibliothek, 513   
1164   Paris, Bibliothèque Nationale de France, fr. 340   
1165        Heidelberg, Universitätsbibliothek, Cpg 397   
1166     Florence, Biblioteca Laurenziana, Plut. 77, 23   

                                             repository     script  \
1     Schloss Anholt (Isselburg), Fürstl. Salm-Salm'...  textualis   
2                              Berlin, Staatsbibliothek        NaN   
3                                    Linz, Landesarchiv        NaN   
4                     Brussels, Koninklijke Bibliotheek  textualis   
5                              Munich, Staatsbibliothek  textualis   
...                                                 ...        ...   
1162                   Orléans, Bibliothèque Municipale        NaN   
1163                   Freiburg, Universitätsbibliothek        NaN   
1164            Paris, Bibliothèque Nationale de France        NaN   
1165                 Heidelberg, Universitätsbibliothek  textualis   
1166                   Florence, Biblioteca Laurenziana        NaN   

      illustrated   surface  text-height  text-width  text-src  illustrations  \
1           False   66220.0        220.0       164.0  original              0   
2           False       NaN          NaN         NaN       NaN              0   
3           False   56000.0        212.0       163.0  original              0   
4           False   35650.0        185.0       132.0  original              0   
5           False   37120.0        195.0       127.0  original              0   
...           ...       ...          ...         ...       ...            ...   
1162        False   15750.0          NaN         NaN       NaN              0   
1163        False   22400.0        150.0       115.0  original              0   
1164         True  126000.0          NaN         NaN       NaN             76   
1165        False   24156.0        165.0       100.0  original              0   
1166        False       NaN          NaN         NaN       NaN              0   

      date_min  date_max  leaf-height  leaf-width  leaf-src  number-lines  \
1       1320.0    1325.0        308.0       215.0   current          32.0   
2       1300.0    1400.0          NaN         NaN       NaN           NaN   
3       1390.0    1410.0        280.0       200.0   current          31.0   
4       1340.0    1360.0        230.0       155.0  original          39.0   
5       1275.0    1300.0        232.0       160.0   current          42.0   
...        ...       ...          ...         ...       ...           ...   
1162    1200.0    1300.0        150.0       105.0   current          30.0   
1163    1290.0    1310.0        160.0       140.0   current           NaN   
1164    1400.0    1425.0        420.0       300.0   current          53.0   
1165    1225.0    1250.0        198.0       122.0   current          26.0   
1166       NaN       NaN          NaN         NaN       NaN           NaN   

       material physical-type               country        region  columns  
1     parchment      fragment                   NaN           NaN      2.0  
2     parchment      fragment               Germany        Bayern      NaN  
3         paper         codex                   NaN           NaN      2.0  
4     parchment      fragment  Southern Netherlands      Flanders      2.0  
5     parchment      fragment                   NaN           NaN      2.0  
...         ...           ...                   ...           ...      ...  
1162  parchment      fragment                   NaN        

### Just Checking...

Let's just see if the first entry by *position* (not index, `iloc` always uses sequential position) has been correctly calculated. This is not exhaustive, but it is a good habit to make sure that things look like they are doing what they should be doing.

In [ ]:
dimensions.iloc[0]

In [ ]:
308 * 215 == df.iloc[0]["surface"]

### More index tricks

Now we'll save the new dataframe we get by accessing the original `df` at the indices from the `dimensions` df, into a new variable called `df_valid`

In [30]:
df_valid = df.loc[dimensions.index]

Sorting is easy in pandas, using `sort_values` (this sorts the whole row, using the values in the column you provide)

In [24]:
df_valid.sort_values(by="surface")

KeyError: 'surface'

... but notice that some of `surface` entries are 0, because one of `leaf-height` or `leaf-width` was mistakenly entered as 0 instead of NA. Let's clean that up.

The main tool for subsetting by conditions inherently uses a boolean Series under the hood, but you can think of it like this: "return the rows of `df_valid` at the indices where `df_valid.surface` (a column) is greater than zero"

In [29]:
df_valid = df_valid[df_valid.surface > 0]
df_valid

AttributeError: 'DataFrame' object has no attribute 'surface'

## Grouping and aggregating

A very powerful method is pandas `groupby()` which creates a special grouping object that can be used for a lot of things. For example, let's see how many illustrated manuscripts occur for each script type...

In [25]:
df_valid.groupby("script")["illustrated"].value_counts()

script               illustrated
cursive              False           77
                     True            25
cursive - hybrida    False            1
cursive - textualis  True             3
hybrida              True            12
                     False            3
hybrida - textualis  True             1
other                False            1
textualis            False          194
                     True           102
Name: count, dtype: int64

We can also "normalize" these numbers by count to obtain comparable proportions

In [ ]:
df_valid.groupby("script")["illustrated"].value_counts(normalize=True)

Finally, and this is just scratching the surface, we can "aggregate" columns from the groups using any function. Common functions (mean, median, std...) can be specific just by their name...

In [31]:
df_valid.groupby("script")["surface"].agg("mean")

script
cursive                 71919.284314
cursive - hybrida       47500.000000
cursive - textualis     54940.000000
hybrida                112492.733333
hybrida - textualis     58212.000000
other                   25024.000000
textualis               66125.557432
Name: surface, dtype: float64

We can subset our valid dataframe again by directly using the boolean column `illustrated` that we created earlier. Now we're down to just 186 manuscripts where we have illustrations as well as valid page sizes.

In [ ]:
df_illustrated = df_valid[df_valid.illustrated]
df_illustrated

### Other formats

Finally, we use pandas again to read the same data from Microsoft Excel.

In [50]:
try:
    import openpyxl
except ImportError:
    %pip install openpyxl

In [ ]:
dfxl = pd.read_excel("../../datasets/arthur/manuscripts.xlsx")
dfxl

Oh no! The Excel indices start from 0, but the CSV we used before started from 1 (maybe it was created in R or Matlab). We can just use our vector operation trick again to add one to the entire index column (ie to each row in it)

In [52]:
dfxl.index += 1

Now (with the indices matching), we use the index locations from our `df_illustrated` dataframe but we grab the rows straight from the new Excel dataframe!

In [ ]:
dfxl.loc[df_illustrated.index]

And finally, a little taste of validation and cleaning. We can use `.equals()` to compare entire `Series` or `DataFrame` objects. Here we find out that the `signature` columns are *not* equal, even though they should be

In [ ]:
dfxl.loc[df_illustrated.index]["signature"].equals(df_illustrated["signature"])

A different method, `.eq()`, returns a boolean `Series` instead (with an equality value for every row). By using a sneaky trick, we assign that to a variable, and then invert it using logical NOT (the tilde `~` operator in pandas). In other words, the original series was `True` almost everywhere and `False` in a couple of places, we swap that, so it is `True` just for the problem rows, and see what indices they are.

Take your time with this idea, it is not simple to understand at first. It can help to make new cells, and then look at individual variables. What is `s`? What does `~s` look like? etc.

In [ ]:
s = dfxl.loc[df_illustrated.index]["signature"].eq(df_illustrated["signature"])
df_illustrated[~s]

... and let's take a look. Oh, this old trick -- someone has accidentally left an extra space at the end of two column entries. This kind of problem comes up all the time in data analysis, and requires constant vigilance. Fixing these problems is called "data cleaning" and,unfortunately, is a necessary process for almost all real-world work.

In [ ]:
dfxl.loc[308]["signature"]

In [ ]:
df_illustrated.loc[308]["signature"]

And to finish, a little trick with Python "f-strings". It is not as complicated as it looks, see if you can figure it out!

(Again, try breaking it into smaller pieces, and then putting them back together!)

In [ ]:
print(
    f"A: {repr(dfxl.loc[596]['signature'])}\nB: {repr(df_illustrated.loc[596]['signature'])}"
)

```
Version History

Current: v1.0.3

10/9/24: 1.0.0: first draft, BN
18/9/24: 1.0.1: add group_by examples, BN
04/10/24: 1.0.2: correct typos and proofread, MK
13/10/24: 1.0.3: move to public, BN
```